src/Database/DbConnection.php

```diff
@@ -25,376 +25,399 @@
+     */
+    protected $connectionName;
+
+    /**
+     * The configuration for the parameters.
+     *
+     * @var array
+     */
+    protected $configuration = [];
+
+    /**
+     * Stores the number of open transactions.
+     *
+     * @var int
+     */
+    protected $transactionCount = 0;
+
+    /**
+     * Stores whether the current transaction has failed.
+     *
+     * @var bool
+     */
+    protected $transactionFailed = false;
+
+    /**
+     * Stores the parameter prefix.
+     *
+     * @var string
+     */
+    protected $paramPrefix = '';
+
+    /**
+     * Opens the connection
+     *
+     * @param array $configuration The configuration for the connection
+     *
+     * @return void
+     */
+    abstract protected function connect(array $configuration);
+
+    /**
+     * Returns the backend type for the given connection {@uses DbFactory::BACKEND_TYPE_*}
+     *
+     * @return string
+     */
+    abstract protected function getBackendType();
+
+
+    /**
+     * Constructor
+     *
+     * @param array  $configuration  The configuration for the parameters.
+     * @param string $connectionName The name of the connection.
+     * @param string $paramPrefix    The prefix for the bound parameters.
+     *
+     * @throws DatabaseException   On connection errors.
+     */
+    public function __construct(array $configuration, $connectionName, $paramPrefix = '')
+    {
+        $this->configuration  = $configuration;
+        $this->connectionName = $connectionName;
+        $this->paramPrefix    = $paramPrefix;
+        try {
+            $this->connect($configuration);
+        } catch (PDOException $exception) {
+            $message = null;
+            $code    = 0;
+            $this->parsePdoException($exception, $message, $code);
+            throw new DatabaseException($message, $code, $exception);
+        }
+        $this->connection->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
+    }
+
+    /**
+     * Closes the database connection.
+     *
+     * @return void
+     */
+    public function disconnect()
+    {
+        $this->connection = null;
+    }
+
+    /**
+     * Returns the prefix which is for prefixing parameters in the query.
+     *
+     * @return string
+     */
+    public function getParamPrefix()
+    {
+        return $this->paramPrefix;
+    }
+
+    /**
+     * Runs a query and returns the result object.
+     *
+     * @param string $query  The query to execute.
+     * @param array  $params The parameters for the query.
+     *
+     * @return \YapepBase\Database\DbResult   The result of the query.
+     *
+     * @throws \YapepBase\Exception\DatabaseException   On execution errors.
+     */
+    public function query($query, array $params = [])
+    {
+        if (empty($this->connection)) {
+            throw new DatabaseException('Connection to the database is not established');
+        }
+
+        $backtrace    = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS,2);
```

Comment:
This will be needed by the debugger. This way we will know from where did the query come from. What I didn't connect yet. 
I was testing it to see if it works and fast and left the working version here as I'll need it later as well.
In the storage branch I refactored the Debugger as well because most of the debugger usage is in the storage, and I refact and test the Dao and DB related stuffs as well here (Not under storage namespace but they are related to the storage too)